### PR TITLE
feat: keys include host-port-suffix

### LINKED ISSUE #1045: Runners getting killed due to runner deployment directory clashes
Steps to reproduce
- `ftl serve --recreate --log-level=DEBUG --idle-runners 3` (default idle runners also hits the issue)
- Wait for idle runners to be loaded, using deployment directories `[...]/ftl-runner/runner[0,1,2]/deployments`
- `ftl deploy examples/go/time -n 10`
- This will cause more runners to be spun up, using deployment directories `[...]/ftl-runner/runner[0,1,2...9]/deployments` (overlaps with above)
- As deployments are reconciled, there'll be errors of `error:runner2: Deployment terminated: signal: killed`, and the deployment count will increase, sometimes decrease, and eventually make its way to 10

### LINKED ISSUE #1047: Update controller and runner ids
In production, use the format `runner-<hostname>-<random>` , `controller-<hostname>-<random>`
Locally use some kind of incrementing id for ease of development

### PR DESCRIPTION
**Pull request by matt2e**
Fixes #1047
Fixes #1045

Changes `ControllerKey` and `RunnerKey` to have the format:
- `<r/c>-<hostname>-<port>-<4bytehex>` for passing around
- `<hostname>-<port>-<4bytehex>` in the db
- non production environments use shorter keys: `r-<incrementing-digits>` and `<incrementing digits>` in db

### CODE REVIEW

File: backend/controller/sql/types.go [hunk 1]
```diff
@@ -1,61 +1,10 @@
...
-	"database/sql/driver"
-	"fmt"
 	"time"
 
 	"github.com/alecthomas/types/optional"
-	"github.com/google/uuid"
-	"github.com/oklog/ulid/v2"
 )
 
-type NullKey = optional.Option[Key]
```
Comment (matt2e) [inline backend/controller/sql/types.go:13]: We have no more uuid/Key columns so this code was all being unused. Is it ok to remove it like I have or are we wanting to keep it for other UUID stuff coming down the road?

File: backend/controller/sql/schema/001_init.sql [hunk 2]
```diff
@@ -101,7 +101,7 @@ CREATE TABLE runners
 (
     id                  BIGINT       NOT NULL GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
     -- Unique identifier for this runner, generated at startup.
-    key                 UUID UNIQUE  NOT NULL,
```
Comment (matt2e) [inline backend/controller/sql/schema/001_init.sql:104]: Are we worried about changing schema without any migration?

File: backend/protos/xyz/block/ftl/v1/ftl.proto [hunk 3]
```diff
@@ -146,7 +146,6 @@ enum RunnerState {
 }
 
 message RegisterRunnerRequest {
-  // UUID representing the runner instance.
```
Comment (matt2e) [inline backend/protos/xyz/block/ftl/v1/ftl.proto:149]: I think this is supposed to update the identifical comment in ftl_pb.ts and ftl.pb.go? Running `bit` didn't seem to trigger anything

File: backend/controller/sql/schema/001_init.sql (hunk 2 above)
Comment (matt2e) [reply backend/controller/sql/schema/001_init.sql:104]: Not at this point

File: backend/controller/scaling/localscaling/local_scaling.go [hunk 4]
```diff
@@ -40,6 +42,7 @@ func NewLocalScaling(portAllocator *bind.BindAllocator, controllerAddresses []*u
 		runners:             map[model.RunnerKey]context.CancelFunc{},
 		portAllocator:       portAllocator,
 		controllerAddresses: controllerAddresses,
+		prevRunnerSuffix:    0, // first runner will have a key of r-0001
```
[the remark this answers is not in the document]
Comment (matt2e) [reply backend/controller/scaling/localscaling/local_scaling.go:45]: Use optional type

[the remark this answers is not in the document]
Comment (matt2e) [reply backend/controller/scaling/localscaling/local_scaling.go:45]: Ok to start at 1

File: backend/controller/sql/types.go (hunk 1 above)
Comment (alecthomas) [reply backend/controller/sql/types.go:13]: Nope nuke it. We can always restore it from history if we need it.

File: internal/model/keys.go [hunk 5]
```diff
@@ -5,82 +5,145 @@ import (
...
+		Suffix:   rand.Intn(10000),
+	}
+}
+func NewLocalRunnerKey(suffix int) RunnerKey {
+	return keyType[runnerKey]{
+		Suffix: suffix,
+	}
+}
+func ParseRunnerKey(key string) (RunnerKey, error)   { return parseKey[RunnerKey](key, true) }
+func ParseRunnerDBKey(key string) (RunnerKey, error) { return parseKey[RunnerKey](key, false) }
```
Comment (alecthomas) [inline internal/model/keys.go:27]: You shouldn't need this function once the DB types are added.

File: internal/model/keys.go [hunk 6]
```diff
@@ -5,82 +5,145 @@ import (
...
+	if d.Hostname == "" {
+		return fmt.Sprintf("%s%04d", prefix, d.Suffix)
+	}
+	return fmt.Sprintf("%s%s-%s-%04d", prefix, d.Hostname, d.Port, d.Suffix)
+}
+
 func (d keyType[T]) MarshalText() ([]byte, error) { return []byte(d.String()), nil }
 func (d *keyType[T]) UnmarshalText(bytes []byte) error {
-	id, err := parseKey[keyType[T]](string(bytes))
+	fmt.Printf("marshal text: %s\n", string(bytes))
```
Comment (alecthomas) [inline internal/model/keys.go:145]: Remove

File: internal/model/keys_test.go [hunk 7]
```diff
@@ -8,9 +8,49 @@ import (
...
+	}{
+		// Production Keys
+		{key: NewRunnerKey("0.0.0.0", "8080"), strPrefix: "r-0.0.0.0-8080-", valuePrefix: "0.0.0.0-8080-"},
+		{key: NewRunnerKey("example-host-with-hyphens", "0"), strPrefix: "r-example-host-with-hyphens-0-", valuePrefix: "example-host-with-hyphens-0-"},
+		{key: NewRunnerKey("noport", ""), strPrefix: "r-noport--", valuePrefix: "noport--"},
+		{key: NewRunnerKey("r-hostwithsameprefix", "80"), strPrefix: "r-r-hostwithsameprefix-80-", valuePrefix: "r-hostwithsameprefix-80-"},
+		{key: NewRunnerKey("r-hostwithprefixandfakeport-80", "80"), strPrefix: "r-r-hostwithprefixandfakeport-80-80-", valuePrefix: "r-hostwithprefixandfakeport-80-80-"},
+
+		// Local Keys
+		{key: NewLocalRunnerKey(0), str: "r-0000", value: "0000"},
```
Comment (alecthomas) [inline internal/model/keys_test.go:26]: Nice

File: internal/model/keys.go [hunk 8]
```diff
@@ -2,85 +2,157 @@
...
+
+func (d keyType[T]) string(includeKind bool) string {
+	var prefix string
+	if includeKind {
+		prefix = fmt.Sprintf("%s-", d.Kind())
+	}
+	if d.Hostname == "" {
+		return fmt.Sprintf("%s%04d", prefix, d.Suffix)
+	}
+	return fmt.Sprintf("%s%s-%s-%04d", prefix, d.Hostname, d.Port, d.Suffix)
```
Comment (alecthomas) [inline internal/model/keys.go:150]: Let's hex encode the suffix with `%08x` so it's consistent with DeploymentName.

File: internal/model/keys.go [hunk 9]
```diff
@@ -2,85 +2,157 @@
...
+}
+func ParseRunnerKey(key string) (RunnerKey, error) { return parseKey[RunnerKey](key, true) }
 
 type runnerKey struct{}
 type RunnerKey = keyType[runnerKey]
 
-func NewControllerKey() ControllerKey                      { return ControllerKey(ulid.Make()) }
-func ParseControllerKey(key string) (ControllerKey, error) { return parseKey[ControllerKey](key) }
+func NewControllerKey(hostname string, port string) ControllerKey {
+	suffix, err := rand.Int(rand.Reader, big.NewInt(10000))
```
Comment (alecthomas) [inline internal/model/keys.go:37]: Let's use 4 bytes (4294967296) of randomness so it's more consistent with DeploymentName.

File: internal/model/keys.go [hunk 10]
```diff
@@ -2,85 +2,157 @@
...
-func parseKey[KT keyType[U], U any](key string) (KT, error) {
-	var zero KT
-	kind := kindFromType[U]()
 	switch {
-	case strings.HasPrefix(key, kind):
-		ulid, err := ulid.Parse(key[len(kind):])
+	case len(components) == 1:
+		//style: [<kind>-]<suffix>
+
+		suffix, err := strconv.Atoi(components[len(components)-1])
```
Comment (alecthomas) [inline internal/model/keys.go:77]: When you switch to hex encoding you can use `strconv.ParseInt()` with a base of 16 here.

**Review by alecthomas — CHANGES_REQUESTED** [review]
LGTM after changing the suffix to more bits of hex-encoded randomness

**Comment by matt2e** [comment]
@alecthomas i switched to 4 bytes of randomness in production keys, moved Suffix to being a string so it can hold hex or incremental ints, and now we can have 0 based indexes again.

**Review by alecthomas — APPROVED** [review]

**Pull request merged**